File: Lab6/lab6_c/lab6.c
#include "lab6.h"

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif


void generate_cities(
    City *cities, 
    int N, 
    double radius, 
    int generate_inside,
    double noise_level
    ) {
        for (int i = 0; i < N; i++) {
            double angle = 2 * M_PI * i / N + ((rand() / (double)RAND_MAX) * noise_level - noise_level / 2);
            double rad = generate_inside ? (rand() / (double)RAND_MAX) * radius : radius;
            cities[i].x = rad * cos(angle);
            cities[i].y = rad * sin(angle);
        }
    }

double **allocate_distance_matrix(int N) {
    double **matrix = malloc(N * sizeof(double *));
    for (int i = 0; i < N; i ++) {
        matrix[i] = malloc(N * sizeof(double));
    }

    return matrix;
}

void free_distance_matrix(double **matrix, int N) {
    for (int i = 0; i < N; i++) {
        free(matrix[i]);
    }
    free(matrix);
}

void calculate_distance_matrix(City *cities, double **dist_mat, int N) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            double dx = cities[i].x - cities[j].x;
            double dy = cities[i].y - cities[j].y;
            dist_mat[i][j] = sqrt(dx * dx + dy * dy);
        }
    }
}

void copy_to_final(int *curr_path, int *final_path, int N) {
    for (int i = 0; i < N; i++) {
        final_path[i] = curr_path[i];
    }
    final_path[N] = curr_path[0];
}

double first_min(double **adj, int i, int N) {
    double min = MAXSIZE;
    for (int j = 0; j < N; j++) {
        if (adj[i][j] < min && i != j) {
            min = adj[i][j];
        }
    }
    return min;
}

double second_min(double **adj, int i, int N) {
    double first = MAXSIZE, second = MAXSIZE;
    for (int j = 0; j < N; j++) {
        if (i == j) continue;
        if (adj[i][j] <= first) {
            second = first;
            first = adj[i][j];
        } else if (adj[i][j] < second) {
            second = adj[i][j];
        }
    }

    return second;
}

void tsp_recursive(double **adj, double curr_bound, double curr_weight, int level, 
                   int *curr_path, int *visited, int *final_path, int N, double *final_res) {
    if (level == N) {
        if (adj[curr_path[level - 1]][curr_path[0]] != 0) {
            double curr_res = curr_weight + adj[curr_path[level - 1]][curr_path[0]];
            if (curr_res < *final_res) {
                copy_to_final(curr_path, final_path, N);
                *final_res = curr_res;
            }
        }
        return;
    }

    for (int i = 0; i < N; i++) {
        if (adj[curr_path[level - 1]][i] != 0 && !visited[i]) {
            double temp = curr_weight + adj[curr_path[level - 1]][i];
            double bound_temp = curr_bound;

            if (level == 1) {
                bound_temp -= (first_min(adj, curr_path[level - 1], N) + first_min(adj, i, N)) / 2;
            } else {
                bound_temp -= (second_min(adj, curr_path[level - 1], N) + first_min(adj, i, N)) / 2;
            }

            if (temp + bound_temp < *final_res) {
                curr_path[level] = i;
                visited[i] = 1;
                tsp_recursive(adj, bound_temp, temp, level + 1, curr_path, visited, final_path, N, final_res);
            }

            visited[i] = 0;
        }
    }
}

void branch_and_bound_tsp(City *cities, int N, int *final_path, double *final_res) {
    double **dist_mat = allocate_distance_matrix(N);
    calculate_distance_matrix(cities, dist_mat, N);

    int curr_path[N + 1], visited[N];
    for (int i = 0; i < N; i++) visited[i] = 0;

    double curr_bound = 0;
    for (int i = 0; i < N; i++) {
        curr_bound += (first_min(dist_mat, i, N) + second_min(dist_mat, i, N));
    }

    curr_bound = ceil(curr_bound / 2);

    curr_path[0] = 0;
    visited[0] = 1;

    *final_res = MAXSIZE;
    tsp_recursive(dist_mat, curr_bound, 0, 1, curr_path, visited, final_path, N, final_res);

    free_distance_matrix(dist_mat, N);
}

int main() {
    int N = 10;
    City cities[N];
    int final_path[N + 1];
    double final_res;

    generate_cities(cities, N, 10.0, 0, 0.1);
    branch_and_bound_tsp(cities, N, final_path, &final_res);

    printf("Minimum cost: %f\n", final_res);
    printf("Path: ");
    for (int i = 0; i <= N; i++) {
        printf("%d ", final_path[i]);
    }
    printf("\n");

    return 0;
}

File: Lab6/lab6_py/lab6.py
import math
import matplotlib.pyplot as plt
import numpy as np
# from mpi4py import MPI
# from mpi4py.futures import MPIPoolExecutor

MAXSIZE = float("inf")

def generate_cities(N: int, radius: float = 10, noise_level: float = 42, generate_inside: bool = False) -> list[tuple[float, float]]:
    angles = np.linspace(0, 2 * np.pi, N, endpoint=False)
    angles += np.random.uniform(-noise_level, noise_level, size=N)
    
    if generate_inside:
        radii = np.random.uniform(0, radius, size=N)
    else:
        radii = np.full(N, radius)

    return [(radii[i] * np.cos(angles[i]), radii[i] * np.sin(angles[i])) for i in range(N)]

def distance_matrix(cities: list[tuple[float, float]]):
    N = len(cities)
    dist_mat: np.ndarray = np.zeros((N, N))

    for i in range(N):
        for j in range(N):
            dist_mat[i][j] = math.sqrt((cities[i][0] - cities[j][0])**2 + (cities[i][1] - cities[j][1])**2)

    return dist_mat

def copy_to_final(curr_path, final_path, N):
    final_path[:N + 1] = curr_path[:]
    final_path[N] = curr_path[0]

def first_min(adj: list[tuple[float, float]], i: int) -> float:
    min = MAXSIZE
    for j in range(len(adj)):
        if adj[i][j] < min and i != j:
            min = adj[i][j]
    return min

def second_min(adj: list[tuple[float, float]], i: int) -> float:
    first = second = MAXSIZE
    for j in range(len(adj)):
        if i == j:
            continue
        if adj[i][j] <= first:
            second = first
            first = adj[i][j]
        elif adj[i][j] <= second and adj[i][j] != first:
            second = adj[i][j]
    return second

def tsp_recursive(adj, curr_bound, curr_weight, level, curr_path, visited, final_path, N, final_res):
    if level == N:
        if adj[curr_path[level - 1]][curr_path[0]] != 0:
            curr_res = curr_weight + adj[curr_path[level - 1]][curr_path[0]]
            if curr_res < final_res[0]:
                copy_to_final(curr_path, final_path, N)
                final_res[0] = curr_res
        return
    
    for i in range(N):
        if adj[curr_path[level-1]][i] != 0 and not visited[i]:
            curr_weight += adj[curr_path[level - 1]][i]
            if level == 1:
                curr_bound -= ((first_min(adj, curr_path[level - 1]) +
                                first_min(adj, i)) / 2)
            else:
                curr_bound -= ((second_min(adj, curr_path[level - 1]) +
                                first_min(adj, i)) / 2)
                
            if curr_bound + curr_weight < final_res[0]:
                visited[i] = True
                curr_path[level] = i
                tsp_recursive(adj, curr_bound, curr_weight, level + 1, curr_path, visited, final_path, N, final_res)
                visited[i] = False

            curr_weight -= adj[curr_path[level - 1]][i]
            curr_bound += ((second_min(adj, curr_path[level - 1]) +
                            first_min(adj, i)) / 2)

def branch_and_bound_tsp(initial_path, N):
    final_res = [MAXSIZE]
    curr_bound = 0
    curr_path = [-1] * (N + 1)
    visited = [False] * N

    dist_mat = distance_matrix(cities)

    for i in range(N):
        curr_bound += (first_min(dist_mat, i) + second_min(dist_mat, i))
    curr_bound = math.ceil(curr_bound / 2)

    curr_path[0] = 0
    visited[0] = True

    final_path = [-1] * (N + 1)
    tsp_recursive(dist_mat, curr_bound, 0, 1, curr_path, visited, final_path, N, final_res)

    return final_res[0], final_path

# def branch_and_bound_tsp_parallel(cities: list[tuple[float, float]], N: int, depth: int):
#     dist_mat = distance_matrix(cities)
#     initial_paths = [[0, i] for i in range(1, min(N, depth+1))]

#     def process_path(path: list[int]) -> tuple[float, list[int]]:
#         visited = [False] * N
#         visited[0] = True
#         curr_bound = 0
#         curr_path = [-1] * (N + 1)
#         curr_path[0] = path[0]
#         for i in range(N):
#             curr_bound += (first_min(dist_mat, i) + second_min(dist_mat, i))
#         curr_bound = math.ceil(curr_bound / 2)

#         final_res = [MAXSIZE]
#         final_path = [-1] * (N + 1)
#         tsp_recursive(dist_mat, curr_bound, 0, 1, curr_path, visited, final_path, N, final_res)
#         return final_res[0], final_path

#     comm = MPI.COMM_WORLD
#     rank = comm.Get_rank()
#     size = comm.Get_size()

#     if rank == 0:
#         initial_paths = comm.bcast(initial_paths, root=0)

#         results = []
#         for i, path in enumerate(initial_paths):
#             result = process_path(path)
#             results.append(result)

#         final_res = [MAXSIZE]
#         final_path = [-1] * (N + 1)
#         for res in results:
#             if res[0] < final_res[0]:
#                 final_res[0] = res[0]
#                 final_path = res[1]

#         return final_res[0], final_path

#     return None, None

def visualize_cities(cities: list[tuple[float, float]]) -> None:
    x_values = [city[0] for city in cities]
    y_values = [city[1] for city in cities]

    plt.figure(figsize=(8, 8))
    plt.scatter(x_values, y_values, color="blue", label="Miasta")

    for i, city in enumerate(cities):
        plt.text(city[0], city[1], str(i), fontsize=12, ha="right")

    plt.title("Rozkład miast")
    plt.xlabel("X")
    plt.ylabel("Y")
    plt.grid(True)
    plt.show()

def visualize_best_path(cities: list[tuple[float, float]], route: list[int]) -> None:
    x_values = [cities[i][0] for i in route]
    y_values = [cities[i][1] for i in route]

    plt.figure(figsize=(8, 8))
    plt.plot(x_values, y_values, color="red", marker="o", linestyle="-", label="Ścieżka")
    plt.scatter(x_values, y_values, color="blue", zorder=5, label="Miasta")

    for i, city in enumerate(cities):
        plt.text(city[0], city[1], str(i), fontsize=12, ha="right")

    plt.title("Optymalna ścieżka")
    plt.xlabel("X")
    plt.ylabel("Y")
    plt.grid(True)
    plt.legend()
    plt.show()

if __name__ == "__main__":
    N = 15
    depth = 2
    cities = generate_cities(N, radius=10, generate_inside=False)
    dist_mat = distance_matrix(cities)

    visualize_cities(cities)

    final_res, final_path = branch_and_bound_tsp(cities, N)

    visualize_best_path(cities, final_path)
    # final_res, best_path = branch_and_bound_tsp_parallel(cities, N, depth)
    # print("Minimum cost:", final_res)
    # print("Path taken:", best_path)
